presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,316 @@
+                // Signal to CBTSF we're done
+                scheduleNextRequest();
+            }
+        }
+    }
+
+    @Override
+    public void fatal(Throwable cause)
+    {
+        try (SetThreadName ignored = new SetThreadName("ContinuousTaskStatusFetcher-%s", idTaskMap)) {
+            updateStats(currentRequestStartNanos.get());
+            onFail.accept(cause);
+        }
+    }
+
+    void updateAllTaskStatus(Map<TaskId, TaskStatus> newValues)
+    {
+        for (Map.Entry<TaskId, TaskStatus> newEntry: newValues.entrySet()) {
+            updateTaskStatus(newEntry.getKey(), newEntry.getValue());
+        }
+    }
+
+    void updateTaskStatus(TaskId taskId, TaskStatus newValue)
+    {
+        StateMachine<TaskStatus> taskStatus = idTaskMap.get(taskId).getValue();
+
+        // change to new value if old value is not changed and new value has a newer version
+        AtomicBoolean taskMismatch = new AtomicBoolean();
+        taskStatus.setIf(newValue, oldValue -> {
+            // did the task instance id change
+            boolean isEmpty = oldValue.getTaskInstanceIdLeastSignificantBits() == 0 && oldValue.getTaskInstanceIdMostSignificantBits() == 0;
+            if (!isEmpty &&
+                    !(oldValue.getTaskInstanceIdLeastSignificantBits() == newValue.getTaskInstanceIdLeastSignificantBits() &&
+                            oldValue.getTaskInstanceIdMostSignificantBits() == newValue.getTaskInstanceIdMostSignificantBits())) {
+                taskMismatch.set(true);
+                return false;
+            }
+
+            if (oldValue.getState().isDone()) {
+                // never update if the task has reached a terminal state
+                return false;
+            }
+            if (newValue.getVersion() < oldValue.getVersion()) {
+                // don't update to an older version (same version is ok)
+                return false;
+            }
+            return true;
+        });
+
+        if (taskMismatch.get()) {
+            // This will also set the task status to FAILED state directly.
+            // Additionally, this will issue a DELETE for the task to the worker.
+            // While sending the DELETE is not required, it is preferred because a task was created by the previous request.
+            onFail.accept(new PrestoException(REMOTE_TASK_MISMATCH, format("%s (%s)", REMOTE_TASK_MISMATCH_ERROR, HostAddress.fromUri(newValue.getSelf()))));
+        }
+    }
+
+    public synchronized boolean isRunning()
+    {
+        return running;
+    }
+
+    /**
+     * Listener is always notified asynchronously using a dedicated notification thread pool so, care should
+     * be taken to avoid leaking {@code this} when adding a listener in a constructor. Additionally, it is
+     * possible notifications are observed out of order due to the asynchronous execution.
+     */
+    public void addStateChangeListener(StateMachine.StateChangeListener<TaskStatus> stateChangeListener)
```

Comment:
we need to call this method from ContinuousTaskStatusFetcher when we add new tasks, and have necessary implementation.